### PR TITLE
Strict type `go_modules`

### PR DESCRIPTION
**Pull request by JamieMagee**
### What are you trying to accomplish?

Complete strict sorbet typing `go_modules`

### Anything you want to highlight for special attention from reviewers?

Also enabled `Sorbet/StrictSigil` to prevent regressions

### How will you know you've accomplished your goal?

Tests pass

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: go_modules/lib/dependabot/go_modules/path_converter.rb [hunk 1]
```diff
@@ -1,19 +1,32 @@
...
+
 require "dependabot/go_modules/native_helpers"
 
 module Dependabot
   module GoModules
     module PathConverter
+      extend T::Sig
+
+      sig do
+        params(path: T.untyped)
```
Comment (jeffwidman) [inline go_modules/lib/dependabot/go_modules/path_converter.rb:14]: Please educate me: why is this untyped and not a string?

File: go_modules/lib/dependabot/go_modules/version.rb [hunk 2]
```diff
@@ -19,41 +21,48 @@ class Version < Dependabot::Version
...
       end
 
+      sig { params(other: Object).returns(T.nilable(Integer)) }
       def <=>(other)
         result = super(other)
         return if result.nil?
         return result unless result.zero?
 
-        other = self.class.new(other) unless other.is_a?(Version)
+        other = self.class.new(other.to_s) unless other.is_a?(Version)
```
Comment (jeffwidman) [inline go_modules/lib/dependabot/go_modules/version.rb:59]: Was this a bug discovered through the typing exercise?

**Review by jeffwidman — APPROVED** [review]
I skimmed this, and it looks fine to me... I didn't take the time to doublecheck every annotation.

Left a question that I trust you probably got it right, and my question is more "help me understand Sorbet in this context"...

File: go_modules/lib/dependabot/go_modules/path_converter.rb (hunk 1 above)
Comment (JamieMagee) [reply go_modules/lib/dependabot/go_modules/path_converter.rb:14]: Good catch! It should be.

File: go_modules/lib/dependabot/go_modules/version.rb (hunk 2 above)
Comment (JamieMagee) [reply go_modules/lib/dependabot/go_modules/version.rb:59]: So the `<=>` method comes from `kernel.rbi` and is defined as

```ruby
sig do
    params(
        other: BasicObject,
    )
    .returns(T.nilable(Integer))
  end
  def <=>(other); end
```

Unfortunately, overriding that directly results in 2 errors:

> Expected `T.nilable(T.any(String, Integer, Gem::Version))` but found `BasicObject` for argument `version`[7002](https://srb.help/7002)

and 

> Method `is_a?` does not exist on `BasicObject` (fix available)[7003](https://srb.help/7003)

I think I could force it to use `VersionParameter` to fix both these, but I am not 100% sure if I can do that. This felt like a compromise.

**Review by amazimbe — APPROVED** [review]

**Pull request merged**